components/mdns/CMakeLists.txt

```
@@ -34,7 +34,26 @@ if(${target} STREQUAL "linux")
     target_link_libraries(${COMPONENT_LIB} PRIVATE "-lbsd")
 endif()
 
-
 if(CONFIG_ETH_ENABLED)
     idf_component_optional_requires(PRIVATE esp_eth)
 endif()
+
+# Get the version number from idf_component.yml
+idf_component_get_property(MDNS_VERSION ${COMPONENT_NAME} COMPONENT_VERSION)
+message(STATUS "Patching mdns.h with version: ${MDNS_VERSION}")
+
+# Specify the original mdns.h file in the "include" folder.
+set(MDNS_H_FILE "${CMAKE_CURRENT_SOURCE_DIR}/include/mdns.h")
+
+# Read the mdns.h file
+file(READ ${MDNS_H_FILE} ORIGINAL_CONTENT)
+
+# Replace only the line that defines ESP_MDNS_VERSION_NUMBER.
+# The regex matches the start of the line with the #define ESP_MDNS_VERSION_NUMBER
+# followed by any characters except a newline.
+string(REGEX REPLACE "#define ESP_MDNS_VERSION_NUMBER[^\n]*"
+       "#define ESP_MDNS_VERSION_NUMBER \"${MDNS_VERSION}\""
+       PATCHED_CONTENT "${ORIGINAL_CONTENT}")
+
+# Write the patched content
+file(WRITE ${MDNS_H_FILE} "${PATCHED_CONTENT}")
```

Have you tried my original suggestion?
```suggestion
target_compile_definitions(${COMPONENT_LIB} PUBLIC "-DESP_MDNS_VERSION_NUMBER=\"${MDNS_VERSION}\"")
```

This should define the macro `ESP_MDNS_VERSION_NUMBER` from compiler command line. The important part is the **`PUBLIC`** keyword, meaning that this wouldn't be defined only in the component itself, but in all **public** dependencies (i.e. everything that includes `mdns.h`). So it should work from the main component.

---

> The key is not propagated to the rest of the files.

it will only be propagated to the component files which define dependency on mdns.

Does the macro expand in the `main` component for you?

> Please, advise how do you want to proceed.

Let's find out why the `target_compile_definitions()` doesn't work for you.
I don't think rewriting the mdns.h header from makefiles (on every expansion?) is acceptable.

---

> Does the macro expand in the main component for you?
In my setup, the macro is not accessable from main().

> I don't think rewriting the mdns.h header from makefiles (on every expansion?) is acceptable.
No problem